cosmwasm/packages/sgx-vm/src/wasmi/exports.rs

```diff
@@ -76,6 +77,59 @@ pub extern "C" fn ocall_read_db(
         .unwrap_or(OcallReturn::Panic)
 }
 
+/// Read a key from the contracts key-value store.
+#[no_mangle]
+pub extern "C" fn ocall_query_chain(
+    context: Ctx,
+    vm_error: *mut UntrustedVmError,
+    gas_used: *mut u64,
+    value: *mut EnclaveBuffer,
+    query: *const u8,
+    query_len: usize,
+) -> OcallReturn {
+    let query = unsafe { std::slice::from_raw_parts(query, query_len) };
+
+    let implementation = unsafe { get_implementations_from_context(&context).query_chain };
+
+    // Returning `EnclaveBuffer { ptr: std::ptr::null_mut() }` is basically returning a null pointer,
+    // which in the enclave is interpreted as signaling that the answer does not exist.
+    // We also interpret this potential panic here as a missing key because we have no way of handling
+    // it at the moment.
+    // In the future, if we see that panics do occur here, we should add a way to report this to the enclave.
+    std::panic::catch_unwind(|| implementation(context, query))
+        // Get either an error(`OcallReturn`), or a response(`EnclaveBuffer`)
+        // which will be converted to a success status.
+        .map(|answer| -> Result<EnclaveBuffer, OcallReturn> {
+            match answer {
+                Ok((Ok(ans), gas_cost)) => {
+                    unsafe { *gas_used = gas_cost };
+                    ans.map(|val| {
+                        super::allocate_enclave_buffer(&val.0).map_err(|_| OcallReturn::Failure)
+                    })
+                    .unwrap_or_else(|_| Ok(EnclaveBuffer::default()))
+                }
+                Ok((Err(_), gas_cost)) => {
+                    unsafe { *gas_used = gas_cost };
+                    Err(OcallReturn::Failure)
+                }
```

Comment:
So, in this context, you shouldn't match on the `SystemResult` as you do here, but rather you need to serialize it and return it into the enclave. The `Ok((Err(_), gas_cost))` is not a failure that this layer handles, but it's a failure that the contract handles.

Reply:
But who consumes the gas?

Reply:
I'm not sure i communicated this correctly. instead of these two cases, you should have:
```rust
Ok((system_result, gas_cost)) => {
  unsafe { *gas_used = gas_cost };
  let response = json_serde::to_vector_or_whatever(system_result).unwrap_or_else(|| EnclaveBuffer::default());
  super::allocate_enclave_buffer(&response).map_err(|_| OcallReturn::Failure)
}
```

Reply:
![c7fc865cf446a4a0d657758ab2ba5f4a](https://user-images.githubusercontent.com/22766533/87443161-ccb47e00-c5fd-11ea-9c46-00b18c2f7349.gif)